OOP Task complete

SiarheiMatsevich opened the pull request:
@akhmialevskiy Please review and take into consideration the following points:

- There was not specified how exactly the 'show_money' method should 'show' money. Based on previous PRs comments - it was agreed that I should not use 'print()' inside methods/functions - so my current implementation only returns (not prints) the amounts. Please let me know if that is the case and if print should be used instead of return, since the 'check_yourself' function looks like it should use print but docstrings saying 'Shows...'.

- The 'check_yourself' function has not been touched, so since the 'join_company' method raises an exception - the execution stops. Please comment out line 235 to prevent it.

- Methods signatures were not touched as well. Only in some places, docstrings have been added. Please let me know if anything else should also be done about it 

The result of execution with commented 235 line and prints added to 'show_money' methods*:

```
Company (Fruits)
Showing money of Alex works at Company (Fruits): 10
Company (Windows and doors)
Showing money of Company (Fruits): 951
Showing money of Company (Windows and doors): 0
Showing money of Alex works at Company (Fruits): 25
Showing money of Bill works at Company (Fruits): 15
Showing money of Jane works at Company (Fruits): 29

Process finished with exit code 0
```
*This is not implemented within the PR and used only for demonstration purposes

SiarheiMatsevich committed:
OOP task completed

SiarheiMatsevich committed:
Fixed Manager's salary

akhmialeuski commented on 3-oop/business.py:
```diff
@@ -0,0 +1,275 @@
+        self.name = name
+        self.address = address
+        self.employees = list()
+        self.__money = 1000
```
I would recommend using constants for magic numbers wherever possible

akhmialeuski commented on 3-oop/business.py:
```diff
@@ -0,0 +1,275 @@
+        if employee in self.employees:
+            # check employee is Engineer
+            if isinstance(employee, Engineer):
+                reward = 10
```
Same, I would move to constant or class attribute

akhmialeuski commented on 3-oop/business.py:
```diff
@@ -0,0 +1,275 @@
+        :rtype: int
+        """
+        # make sure engineer is employed to this company
+        if employee in self.employees:
```
You have `is_employed`

akhmialeuski commented on 3-oop/business.py:
```diff
@@ -0,0 +1,275 @@
+                reward = 5
+                self.__money -= reward
+                employee.bonus_to_salary(company=self, reward=reward)
+            if self.is_bankrupt:
```
What if the company goes bankrupt while the bonuses are being paid out? Will it go into negative equity and then go bankrupt?

akhmialeuski commented on 3-oop/business.py:
```diff
@@ -0,0 +1,275 @@
+        """
+        Company should call this method on each employee when having a party
+        """
+        if company == self.company:
```
I don't understand the meaning of this checking, please explain it to me

akhmialeuski commented on 3-oop/business.py:
```diff
@@ -0,0 +1,275 @@
+        else:
+            raise ValueError('Employee is not an Engineer or Manager')
+
+    def dismiss_employee(self, employee):
```
Looks like you don't use this method

akhmialeuski commented:
- Please keep a consistent style in docstrings and variable descriptions. This can be configured once in the IDE and follow this style. It does not concern the task itself, but you should habituate yourself to a good programming style right away
- This isn't directly related to the task itself, but I would ask that you use type hinting in functions throughout your code

akhmialeuski commented on 3-oop/business.py:
```diff
@@ -0,0 +1,275 @@
+    bill.join_company(fruits_company)
+
+    # doors becomes a bankrupt
+    doors_company.go_bankrupt()
```
It work wrong

```python
# doors becomes a bankrupt
    doors_company.go_bankrupt()
    alex.join_company(doors_company)
    alex.do_work()
    alex.do_work()
    alex.do_work()
    print(alex.show_money())
    print(doors_company.show_money())
```
got
```
Company (Fruits)
Company (Windows and doors)
50
-30
50
15
29
```

SiarheiMatsevich replied:
```
newc.go_bankrupt()
jane.join_company(newc)
jane.do_work()
Traceback (most recent call last):
  File "C:\Program Files\JetBrains\PyCharm Community Edition 2023.2\plugins\python-ce\helpers\pydev\pydevconsole.py", line 364, in runcode
    coro = func()
           ^^^^^^
  File "<input>", line 1, in <module>
  File "D:\PycharmProjects\python-automation-mentoring\business.py", line 256, in do_work
    self.put_money_into_my_wallet(self.company.write_reports(self))
                                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "D:\PycharmProjects\python-automation-mentoring\business.py", line 96, in write_reports
    raise ValueError(f'The {self} has not enough money to pay for the work')
ValueError: The Company (newC) has not enough money to pay for the work
```

akhmialeuski commented on 3-oop/business.py:
```diff
@@ -0,0 +1,275 @@
+    jane.join_company(fruits_company)
+    # Jane works pretty hard. She writes lots of reports
+    jane.do_work()
+    jane.do_work()
```
```python
jane = Manager('Jane', 30)
    jane.join_company(fruits_company)
    # Jane works pretty hard. She writes lots of reports
    for _ in range(1000):
        jane.do_work()
        jane.do_work()
        jane.do_work()

    # # results
    print(f'fruits_company: {fruits_company.show_money()}')

    print(f'alex: {alex.show_money()}')
    print(f'bill: {bill.show_money()}')
    print(f'jane: {jane.show_money()}')
```

got

```
Company (Fruits)
Company (Windows and doors)
fruits_company: -35010
alex: 20
bill: 10
jane: 36000
```

akhmialeuski replied:
It work wrong

SiarheiMatsevich replied:
I tried the same and got an exception:
```
newc = business.Company('newC', 'Address')
newc.add_employee(jane)
for i in range(1000):
    jane.do_work()
    jane.do_work()
    jane.do_work()
    
Traceback (most recent call last):
  File "C:\Program Files\JetBrains\PyCharm Community Edition 2023.2\plugins\python-ce\helpers\pydev\pydevconsole.py", line 364, in runcode
    coro = func()
           ^^^^^^
  File "<input>", line 4, in <module>
  File "D:\PycharmProjects\python-automation-mentoring\business.py", line 256, in do_work
    self.put_money_into_my_wallet(self.company.write_reports(self))
                                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "D:\PycharmProjects\python-automation-mentoring\business.py", line 96, in write_reports
    raise ValueError(f'The {self} has not enough money to pay for the work')
ValueError: The Company (newC) has not enough money to pay for the work
```

akhmialeuski commented:
Please test your code properly before submitting it for review

SiarheiMatsevich committed:
Changes:
- Fixed add_employee method. Now a company can employ a person as well as an employee can use join_company method to join a company
- Fixed do_tasks and write_reports method. Now those methods check if a company has enough money to pay for a job and raise an exception if it has not enough money and call go_bankrupt if company balance becomes 0
- Fixed __repr__ method of Employee class
- Formatted all docstrings
- Added hints to methods
